File: commons/src/main/java/ir/msob/jima/core/commons/security/BaseUser.java
package ir.msob.jima.core.commons.security;

import ir.msob.jima.core.commons.shared.ModelType;
import lombok.*;

import java.io.Serial;
import java.util.SortedSet;
import java.util.TreeSet;

/**
 * This class represents a base user with an ID, session ID, username, roles, and audience.
 * It extends the ModelType class and implements Serializable interface for the object to be converted into a byte stream.
 * It provides methods to get and set the ID, session ID, username, roles, and audience.
 */
@Setter
@Getter
@ToString(callSuper = true)
@NoArgsConstructor
@AllArgsConstructor
public class BaseUser extends ModelType {
    /**
     * The serial version UID for the serialization.
     */
    @Serial
    private static final long serialVersionUID = 4945340068393543863L;

    /**
     * The ID of the user.
     */
    private String id;

    /**
     * The session ID of the user.
     */
    private String sessionId;

    /**
     * The username of the user.
     */
    private String name;

    /**
     * The username of the user.
     */
    private String username;

    /**
     * The roles of the user.
     */
    private SortedSet<String> roles = new TreeSet<>();

    /**
     * The audience of the user.
     */
    private String audience;
}

File: web/angular/projects/core-commons/src/lib/model/security/base-user.ts
import {ModelType} from "../dto/model-type";

export class BaseUser extends ModelType {
  private _id: string | undefined;

  get id(): string | undefined {
    return this._id;
  }

  set id(id: string | undefined) {
    this._id = id;
  }

  private _sessionId: string | undefined;

  get sessionId(): string | undefined {
    return this._sessionId;
  }

  set sessionId(sessionId: string | undefined) {
    this._sessionId = sessionId;
  }
  private _name: string | undefined;

  get name(): string | undefined {
    return this._name;
  }

  set name(value: string | undefined) {
    this._name = value;
  }

  private _username: string | undefined;

  get username(): string | undefined {
    return this._username;
  }

  set username(username: string | undefined) {
    this._username = username;
  }

  private _roles: string[] = [];

  get roles(): string[] {
    return this._roles;
  }

  set roles(roles: string[]) {
    this._roles = roles;
  }

  private _audience: string | undefined;

  get audience(): string | undefined {
    return this._audience;
  }

  set audience(audience: string | undefined) {
    this._audience = audience;
  }
}
